subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java

```diff
@@ -60,16 +68,31 @@ public void waitForCompletion(Operation operation) {
         }
 
         try {
-            for (AsyncWorkCompletion item : workItems) {
-                try {
-                    item.waitForCompletion();
-                } catch (Throwable t) {
-                    failures.add(t);
-                }
-            }
+            if (workItems.size() > 0) {
+                boolean workInProgress = CollectionUtils.any(workItems, new Spec<AsyncWorkCompletion>() {
+                    @Override
+                    public boolean isSatisfiedBy(AsyncWorkCompletion workCompletion) {
+                        return !workCompletion.isComplete();
+                    }
+                });
+                if (workInProgress) {
```

Comment:
I think the concern is legitimate.  I think our options are:

1. Don't wait for work in between task actions, just wait until all task actions have run.  This would mean that task actions could not have any dependency on each other, which seems like it could inhibit usage of the new api.
1. Wait for work in between task actions, but only release the project lock after the last task action.  This would mean less opportunities for parallelism but would eliminate the potential for project state changing in between task actions.
1. Keep it the way it is, but promote the practice that actions should avoid relying on project state to be exactly the same for all actions (i.e. we can guarantee that project state won't change while a given task action is executing, but not in between actions, or when the user waits on async work).
1. Same as previous, but don't release the project while the user is waiting on work (i.e. only in between task actions).  This would just provide the added guarantee that the project state won't change at all during the execution of a task action (even if the user waits on work).

It seems like you can imagine scenarios where any of these could be problematic.  My gut feeling is that what we have right now is probably fine, in that I expect the vast majority of task actions will simply kick off work and exit and if there are multiple waits during task execution the later work will tend to rely on the results of the earlier work rather than on (potentially modified) project state.